diff --git a/controller/lport.c b/controller/lport.c
index 9cb7308ce..5371e6745 100644
--- a/controller/lport.c
+++ b/controller/lport.c
@@ -175,0 +176,4 @@ lport_can_bind_on_this_chassis(const struct sbrec_chassis *chassis_rec,
+        if (pb->n_requested_additional_chassis == 1 &&
+            pb->requested_additional_chassis[0] == chassis_rec) {
+                return CAN_BIND_AS_ADDITIONAL;
+        }
