diff --git a/northd/northd.c b/northd/northd.c
index c5121e0b5..764dbd596 100644
--- a/northd/northd.c
+++ b/northd/northd.c
@@ -9332,0 +9333,68 @@ build_lswitch_dnat_mod_dl_dst_rules(struct ovn_port *op,
+static void
+build_lswitch_mod_dl_src_rules(struct ovn_port *op,
+                               struct lflow_table *lflows,
+                               const struct hmap *lr_ports,
+                               struct ds *actions,
+                               struct ds *match)
+{
+    if (!op->nbsp || !op->od || !op->od->nbs ||
+        !op->od->n_router_ports || !op->od->n_localnet_ports) {
+        return;
+    }
+    if (!lsp_is_enabled(op->nbsp)) {
+        return;
+    }
+    if (!strcmp(op->nbsp->type, "virtual") ||
+        !strcmp(op->nbsp->type, "localport")) {
+        return;
+    }
+    if (lsp_is_external(op->nbsp) || lsp_is_router(op->nbsp) ||
+        op->has_unknown) {
+        return;
+    }
+
+    if (op->n_lsp_addrs != 1 || !strlen(op->lsp_addrs[0].ea_s) ||
+        (!op->lsp_addrs[0].n_ipv4_addrs && !op->lsp_addrs[0].n_ipv6_addrs)) {
+        return;
+    }
+
+    ds_clear(actions);
+    ds_put_format(actions, "eth.src = %s; next(pipeline=ingress, table=%d);",
+                  op->lsp_addrs[0].ea_s,
+                  ovn_stage_get_table(S_SWITCH_IN_L2_LKUP));
+
+    for (size_t i = 0; i < op->od->n_router_ports; i++) {
+        struct ovn_port *rp = op->od->router_ports[i];
+        if (!rp || !rp->nbsp) {
+            continue;
+        }
+        struct ovn_port *peer = ovn_port_get_peer(lr_ports, rp);
+        if (!peer || !peer->nbrp || peer->primary_port ||
+            peer->nbrp->n_gateway_chassis || peer->nbrp->ha_chassis_group) {
+            continue;
+        }
+
+        for (size_t j = 0; j < op->lsp_addrs[0].n_ipv4_addrs; j++) {
+            ds_clear(match);
+            ds_put_format(match,
+                          "inport == %s && ip4.src == %s && eth.src != %s",
+                          rp->json_key, op->lsp_addrs[0].ipv4_addrs[j].addr_s,
+                          op->lsp_addrs[0].ea_s);
+            ovn_lflow_add_with_hint(lflows, op->od, S_SWITCH_IN_L2_LKUP, 60,
+                                    ds_cstr(match), ds_cstr(actions),
+                                    &op->nbsp->header_, op->lflow_ref);
+        }
+
+        for (size_t j = 0; j < op->lsp_addrs[0].n_ipv6_addrs; j++) {
+            ds_clear(match);
+            ds_put_format(match,
+                          "inport == %s && ip6.src == %s && eth.src != %s",
+                          rp->json_key, op->lsp_addrs[0].ipv6_addrs[j].addr_s,
+                          op->lsp_addrs[0].ea_s);
+            ovn_lflow_add_with_hint(lflows, op->od, S_SWITCH_IN_L2_LKUP, 60,
+                                    ds_cstr(match), ds_cstr(actions),
+                                    &op->nbsp->header_, op->lflow_ref);
+        }
+    }
+}
+
@@ -20046,0 +20115 @@ build_lswitch_and_lrouter_iterate_by_lsp(struct ovn_port *op,
+    build_lswitch_mod_dl_src_rules(op, lflows, lr_ports, actions, match);
