From 71f831b9cc5a6dc923af4ca90286857e2cf8b1d3 Mon Sep 17 00:00:00 2001
From: Mengxin Liu <mengxin@alauda.io>
Date: Wed, 9 Feb 2022 18:11:55 +0800
Subject: [PATCH] change hash type from dp_hash to hash with field src_ip

---
 lib/actions.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/lib/actions.c b/lib/actions.c
index b3433f49ea..7ead4a673e 100644
--- a/lib/actions.c
+++ b/lib/actions.c
@@ -1357,7 +1357,7 @@ encode_SELECT(const struct ovnact_select *select,
     struct ofpact_group *og;
 
     struct ds ds = DS_EMPTY_INITIALIZER;
-    ds_put_format(&ds, "type=select,selection_method=dp_hash");
+    ds_put_format(&ds, "type=select,selection_method=hash,fields=ip_src");
 
     struct mf_subfield sf = expr_resolve_field(&select->res_field);
 
