diff --git a/controller/lflow.c b/controller/lflow.c
index 382c2aecb..0fcde2ab5 100644
--- a/controller/lflow.c
+++ b/controller/lflow.c
@@ -35,0 +36 @@
+#include "lib/ovn-util.h"
@@ -1094,0 +1096,10 @@ consider_logical_flow__(const struct sbrec_logical_flow *lflow,
+    if (!lflow_kube_ovn_skip_ct) {
+        const char *hint = smap_get(&lflow->external_ids, "kube-ovn-hint");
+        if (hint && !strcmp(hint, OVN_LFLOW_HINT_KUBE_OVN_SKIP_CT)) {
+            VLOG_DBG("lflow "UUID_FMT
+                      " is not compatible with current kernel version, skip",
+                      UUID_ARGS(&lflow->header_.uuid));
+            return;
+        }
+    }
+
diff --git a/controller/ovn-controller.h b/controller/ovn-controller.h
index 57d5577bc..cb255e7a2 100644
--- a/controller/ovn-controller.h
+++ b/controller/ovn-controller.h
@@ -20,0 +21 @@
+#include <stdbool.h>
@@ -28,0 +30,2 @@ enum chassis_tunnel_type get_tunnel_type(const char *name);
+extern bool lflow_kube_ovn_skip_ct;
+
diff --git a/lib/ovn-util.h b/lib/ovn-util.h
index 4ef93c470..280eb5024 100644
--- a/lib/ovn-util.h
+++ b/lib/ovn-util.h
@@ -36,0 +37,2 @@
+#define OVN_LFLOW_HINT_KUBE_OVN_SKIP_CT "kube-ovn: skip conntrack"
+
