From 9ee66bd91be65605cffb9a490b4dba3bc13358e9 Mon Sep 17 00:00:00 2001
From: clyi <clyi@alauda.io>
Date: Tue, 5 Aug 2025 10:30:31 +0800
Subject: [PATCH] modify src route priority

Signed-off-by: clyi <clyi@alauda.io>
---
 northd/northd.c | 10 +++++++++-
 1 file changed, 9 insertions(+), 1 deletion(-)

diff --git a/northd/northd.c b/northd/northd.c
index cf9d2217ae..c530468ec2 100644
--- a/northd/northd.c
+++ b/northd/northd.c
@@ -11455,7 +11455,15 @@ build_route_match(const struct ovn_port *op_inport, uint32_t rtb_id,
     if (has_protocol_match) {
         ofs += 1;
     }
-    *priority = (plen * ROUTE_PRIO_OFFSET_MULTIPLIER) + ofs;
+
+    if (is_src_route) {
+        *priority = 1;
+        if ((is_ipv4 && plen == 32) || (plen == 128)) {
+            *priority = ROUTE_PRIO_OFFSET_STATIC + 1;
+        }
+    } else {
+        *priority = (plen * ROUTE_PRIO_OFFSET_MULTIPLIER) + ofs;
+    }
 
     ds_put_format(match, "ip%s.%s == %s/%d", is_ipv4 ? "4" : "6", dir,
                   network_s, plen);
