From c7b1a0f0675f31aee9cf0250a1294527de29eafb Mon Sep 17 00:00:00 2001
From: zhangzujian <zhangzujian.7@gmail.com>
Date: Thu, 10 Apr 2025 01:29:00 +0000
Subject: [PATCH] modify src route priority

Signed-off-by: zhangzujian <zhangzujian.7@gmail.com>
---
 northd/northd.c | 9 +++++++++
 1 file changed, 9 insertions(+)

diff --git a/northd/northd.c b/northd/northd.c
index cf9d2217ae..882f9475f8 100644
--- a/northd/northd.c
+++ b/northd/northd.c
@@ -11444,6 +11444,15 @@ build_route_match(const struct ovn_port *op_inport, uint32_t rtb_id,
         dir = "dst";
     }
 
+    if (is_src_route) {
+        *priority = 1;
+        if ((is_ipv4 && plen == 32) || (plen == 128)) {
+            *priority = 2;
+        }
+    } else {
+        *priority = (plen * ROUTE_PRIO_OFFSET_MULTIPLIER) + ofs;
+    }
+
     if (op_inport) {
         ds_put_format(match, "inport == %s && ", op_inport->json_key);
     }
