From e490f5ac0b644101913c2a3db8e03d85e859deff Mon Sep 17 00:00:00 2001
From: Mengxin Liu <mengxin@alauda.io>
Date: Thu, 10 Apr 2025 01:28:59 +0000
Subject: [PATCH] change default hash type from dp_hash to hash with field
 src_ip
---
 lib/actions.c | 8 +++++---
 1 file changed, 5 insertions(+), 3 deletions(-)

diff --git a/lib/actions.c b/lib/actions.c
index 26df8e203..3f6572445 100644
--- a/lib/actions.c
+++ b/lib/actions.c
@@ -1729,2 +1728,0 @@ encode_SELECT(const struct ovnact_select *select,
-    ds_put_format(&ds, "type=select,selection_method=%s",
-                  select->hash_fields ? "hash": "dp_hash");
@@ -1732 +1730,4 @@ encode_SELECT(const struct ovnact_select *select,
-      ds_put_format(&ds, ",fields(%s)", select->hash_fields);
+        ds_put_format(&ds, "type=select,selection_method=hash,fields(%s)",
+                      select->hash_fields);
+    } else {
+        ds_put_format(&ds, "type=select,selection_method=hash,fields=ip_src");
