From d088c5d8c263552c5a31d87813991aee30ab74de Mon Sep 17 00:00:00 2001
From: zhangzujian <zhangzujian.7@gmail.com>
Date: Tue, 26 Nov 2024 16:29:08 +0800
Subject: [PATCH] ovs-router: skip getting source address for kube-ipvs0

Signed-off-by: zhangzujian <zhangzujian.7@gmail.com>
---
 lib/ovs-router.c | 6 +++++-
 1 file changed, 5 insertions(+), 1 deletion(-)

diff --git a/lib/ovs-router.c b/lib/ovs-router.c
index 2566386ea2..fcdc3072a3 100644
--- a/lib/ovs-router.c
+++ b/lib/ovs-router.c
@@ -218 +217,0 @@ ovs_router_lookup(uint32_t mark, const struct in6_addr *ip6_dst,
-
@@ -335,0 +335,5 @@ ovs_router_get_netdev_source_address(const struct in6_addr *ip6_dst,
+    if (!strcmp(netdev_name, "kube-ipvs0")) {
+        VLOG_DBG_RL(&rl, "skip getting source address for device kube-ipvs0");
+        return 0;
+    }
+
