diff --git a/controller/physical.c b/controller/physical.c
index 030010c33..effc5f7c9 100644
--- a/controller/physical.c
+++ b/controller/physical.c
@@ -44,0 +45 @@
+#include "lib/ovn-l7.h"
@@ -2313,0 +2315,30 @@ consider_port_binding(const struct physical_ctx *ctx,
+        if (smap_get_bool(&binding->options, "bfd-only", false)) {
+            match_set_nw_proto(&match, IPPROTO_UDP);
+            match_set_tp_dst(&match, htons(BFD_DEST_PORT));
+            ofpbuf_clear(ofpacts_p);
+            encode_controller_op(ACTION_OPCODE_BFD_MSG, 0, ofpacts_p);
+
+            for (size_t i = 0; i < binding->n_mac; i++) {
+                struct lport_addresses laddrs;
+                if (!extract_lsp_addresses(binding->mac[i], &laddrs)) {
+                    continue;
+                }
+
+                for (size_t j = 0; j < laddrs.n_ipv4_addrs; j++) {
+                    match_set_nw_dst(&match, laddrs.ipv4_addrs[j].addr);
+                    ofctrl_add_flow(flow_table, OFTABLE_LOCAL_OUTPUT, 110,
+                        binding->header_.uuid.parts[0],
+                        &match, ofpacts_p, &binding->header_.uuid);
+                }
+                match_set_nw_dst(&match, 0);
+
+                for (size_t j = 0; j < laddrs.n_ipv6_addrs; j++) {
+                    match_set_ipv6_dst(&match, &laddrs.ipv6_addrs[j].addr);
+                    ofctrl_add_flow(flow_table, OFTABLE_LOCAL_OUTPUT, 110,
+                                    binding->header_.uuid.parts[0],
+                                    &match, ofpacts_p, &binding->header_.uuid);
+                }
+
+                destroy_lport_addresses(&laddrs);
+            }
+        }
diff --git a/controller/pinctrl.c b/controller/pinctrl.c
index 3b7deba48..bafe2b22f 100644
--- a/controller/pinctrl.c
+++ b/controller/pinctrl.c
@@ -7956,0 +7957 @@ bfd_monitor_run(struct ovsdb_idl_txn *ovnsb_idl_txn,
+        bool bfd_only = smap_get_bool(&pb->options, "bfd-only", false);
@@ -7958 +7959 @@ bfd_monitor_run(struct ovsdb_idl_txn *ovnsb_idl_txn,
-        if (!peer_s) {
+        if (!peer_s && !bfd_only) {
@@ -7962,4 +7963,6 @@ bfd_monitor_run(struct ovsdb_idl_txn *ovnsb_idl_txn,
-        const struct sbrec_port_binding *peer
-            = lport_lookup_by_name(sbrec_port_binding_by_name, peer_s);
-        if (!peer) {
-            continue;
+        if (peer_s && !bfd_only) {
+            const struct sbrec_port_binding *peer
+                = lport_lookup_by_name(sbrec_port_binding_by_name, peer_s);
+            if (!peer) {
+                continue;
+            }
diff --git a/northd/northd.c b/northd/northd.c
index 4d2fbfaa7..186438210 100644
--- a/northd/northd.c
+++ b/northd/northd.c
@@ -4099,0 +4100,5 @@ sync_pb_for_lrp(struct ovn_port *op,
+    const bool bfd_only = smap_get_bool(&op->nbrp->options, "bfd-only", false);
+    if (bfd_only) {
+        smap_add(&new, "bfd-only", "true");
+    }
+
@@ -9140 +9145 @@ build_lswitch_dnat_mod_dl_dst_rules(struct ovn_port *op,
-                                    const struct hmap *lr_ports,
+                                    const struct hmap *lr_ports OVS_UNUSED,
@@ -11954,4 +11959,7 @@ find_policy_outport(struct ovn_datapath *od, const struct hmap *lr_ports,
-static bool check_bfd_state(const struct nbrec_logical_router_policy *rule,
-                            struct ovn_port *out_port, const char *nexthop,
-                            const struct hmap *bfd_connections,
-                            struct hmap *bfd_active_connections)
+static bool check_bfd_state(
+        const struct nbrec_logical_router_policy *rule,
+        const struct hmap *lr_ports,
+        const struct hmap *bfd_connections,
+        struct ovn_port *out_port,
+        const char *nexthop,
+        struct hmap *bfd_active_connections)
@@ -11979 +11987,5 @@ static bool check_bfd_state(const struct nbrec_logical_router_policy *rule,
-            continue;
+            struct ovn_port *op = ovn_port_find(lr_ports, nb_bt->logical_port);
+            if (!op || !op->nbrp ||
+                !smap_get_bool(&op->nbrp->options, "bfd-only", false)) {
+                continue;
+            }
@@ -12035,0 +12048 @@ build_routing_policy_flow(struct lflow_table *lflows, struct ovn_datapath *od,
+
@@ -12128,0 +12142 @@ build_ecmp_routing_policy_flows(struct lflow_table *lflows,
+
@@ -14657,0 +14672,3 @@ build_lrouter_bfd_flows(struct lflow_table *lflows, struct ovn_port *op,
+    char *redirect_name = ovn_chassis_redirect_name(op->nbrp->name);
+    char *actions = xasprintf("outport = \"%s\"; output;", redirect_name);
+    bool bfd_only = smap_get_bool(&op->nbrp->options, "bfd-only", false);
@@ -14674,0 +14692,14 @@ build_lrouter_bfd_flows(struct lflow_table *lflows, struct ovn_port *op,
+        if ((op->nbrp->ha_chassis_group || op->nbrp->n_gateway_chassis) &&
+            bfd_only) {
+            ds_clear(&match);
+            ds_put_format(&match, "ip4.dst == %s && udp.dst == 3784 && "
+                          "!is_chassis_resident(\"%s\")",
+                          ds_cstr(&ip_list), redirect_name);
+            ovn_lflow_add_with_hint__(lflows, op->od, S_ROUTER_IN_IP_INPUT,
+                                      115, ds_cstr(&match), actions, NULL,
+                                      copp_meter_get(COPP_BFD,
+                                                     op->od->nbr->copp,
+                                                     meter_groups),
+                                      &op->nbrp->header_,
+                                      lflow_ref);
+        }
@@ -14694,0 +14727,14 @@ build_lrouter_bfd_flows(struct lflow_table *lflows, struct ovn_port *op,
+        if ((op->nbrp->ha_chassis_group || op->nbrp->n_gateway_chassis) &&
+            bfd_only) {
+            ds_clear(&match);
+            ds_put_format(&match, "ip6.dst == %s && udp.dst == 3784 && "
+                          "!is_chassis_resident(\"%s\")",
+                          ds_cstr(&ip_list), redirect_name);
+            ovn_lflow_add_with_hint__(lflows, op->od, S_ROUTER_IN_IP_INPUT,
+                                      115, ds_cstr(&match), actions, NULL,
+                                      copp_meter_get(COPP_BFD,
+                                                     op->od->nbr->copp,
+                                                     meter_groups),
+                                      &op->nbrp->header_,
+                                      lflow_ref);
+        }
@@ -14698,0 +14746,2 @@ build_lrouter_bfd_flows(struct lflow_table *lflows, struct ovn_port *op,
+    free(redirect_name);
+    free(actions);
@@ -15652,2 +15701,2 @@ build_route_policies(struct ovn_datapath *od, const struct hmap *lr_ports,
-                if (!check_bfd_state(rule, out_port, nexthop,
-                                     bfd_connections,
+                if (!check_bfd_state(rule, lr_ports, bfd_connections,
+                                     out_port, nexthop,
