From e5916eb53abc3b7d28c407c3c47566c46116090a Mon Sep 17 00:00:00 2001
From: zhangzujian <zhangzujian.7@gmail.com>
Date: Thu, 10 Apr 2025 01:31:33 +0000
Subject: [PATCH] fix lr-lb dnat with multiple distributed gateway ports

Signed-off-by: zcq98 <zhaocongqi_yewu@cmss.chinamobile.com>
Signed-off-by: zhangzujian <zhangzujian.7@gmail.com>
---
 northd/northd.c | 16 ++++++++++++++--
 1 file changed, 14 insertions(+), 2 deletions(-)

diff --git a/northd/northd.c b/northd/northd.c
index 6de9968f9..4d2fbfaa7 100644
--- a/northd/northd.c
+++ b/northd/northd.c
@@ -13386,0 +13387,14 @@ build_distr_lrouter_nat_flows_for_lb(struct lrouter_nat_lb_flows_ctx *ctx,
+        if (vector_len(&od->l3dgw_ports) > 1) {
+            dgp = NULL;
+            for (size_t i = 0; i < vector_len(&od->l3dgw_ports); i++) {
+                struct ovn_port *candidate = vector_get(&od->l3dgw_ports, i,
+                                                        struct ovn_port *);
+                if (lrp_find_member_ip(candidate, ctx->lb_vip->vip_str)) {
+                    dgp = candidate;
+                    break;
+                }
+            }
+            if (!dgp) {
+                return;
+            }
+        }
