diff --git a/northd/en-global-config.c b/northd/en-global-config.c
index afc559160..771508903 100644
--- a/northd/en-global-config.c
+++ b/northd/en-global-config.c
@@ -723,0 +724,6 @@ check_nb_options_out_of_sync(
+                           "acl_ct_translation", false)) {
+        return true;
+    }
+
+    if (config_out_of_sync(&nb->options, &config_data->nb_options,
+                           "bcast_arp_req_flood", false)) {
diff --git a/northd/northd.c b/northd/northd.c
index e6fc96b8b..8e3e62a02 100644
--- a/northd/northd.c
+++ b/northd/northd.c
@@ -106,0 +107,5 @@ static bool vxlan_mode;
+static bool vxlan_ic_mode;
+
+static bool bcast_arp_req_flood = true;
+
+
@@ -10145,0 +10151,5 @@ build_lswitch_lflows_l2_unknown(struct ovn_datapath *od,
+        if (!bcast_arp_req_flood) {
+            ovn_lflow_add(lflows, od, S_SWITCH_IN_L2_LKUP, 90,
+                          "eth.bcast && arp.op == 1", "next;",
+                          lflow_ref);
+        }
@@ -10660,0 +10671,44 @@ build_lswitch_arp_nd_responder_default(struct ovn_datapath *od,
+/* Ingress table 24: ARP/ND responder for service monitor source ip.
+/* Ingress table 19: forward ARP/ND request from localnet/vtep
+ * to logical switch port for known IPs.
+ * (priority 110)*/
+static void
+build_lswitch_arp_nd_forward_for_unknown_ips(struct ovn_port *op,
+                                             struct lflow_table *lflows,
+                                             struct ds *actions,
+                                             struct ds *match)
+{
+    if (bcast_arp_req_flood) {
+        return;
+    }
+    if (!op->od->nbs) {
+        return;
+    }
+    if (is_vlan_transparent(op->od)) {
+        return;
+    }
+
+    for (size_t i = 0; i < op->od->n_localnet_ports; i++) {
+        struct ovn_port *localnet_port = op->od->localnet_ports[i];
+        if (!localnet_port->nbsp) {
+            continue;
+        }
+
+        for (size_t j = 0; j < op->n_lsp_addrs; j++) {
+            for (size_t k = 0; k < op->lsp_addrs[j].n_ipv4_addrs; k++) {
+                ds_clear(match);
+                ds_put_format(match,
+                              "inport == %s && eth.bcast && "
+                              "arp.op == 1 && arp.tpa == %s",
+                              localnet_port->json_key,
+                              op->lsp_addrs[j].ipv4_addrs[k].addr_s);
+                ds_clear(actions);
+                ds_put_format(actions, "outport = %s; output;",
+                              op->json_key);
+                ovn_lflow_add(lflows, op->od, S_SWITCH_IN_ARP_ND_RSP, 110,
+                              ds_cstr(match), ds_cstr(actions), op->lflow_ref);
+            }
+        }
+    }
+}
+
@@ -19485,0 +19540 @@ build_lswitch_and_lrouter_iterate_by_lsp(struct ovn_port *op,
+    build_lswitch_arp_nd_forward_for_unknown_ips(op, lflows, actions, match);
@@ -20969,0 +21025,2 @@ ovnnb_db_run(struct northd_input *input_data,
+    bcast_arp_req_flood = smap_get_bool(input_data->nb_options,
+                                        "bcast_arp_req_flood", true);
