From 54056ea65dc28aa1c4c721a2a34d7913f79f8376 Mon Sep 17 00:00:00 2001
From: Mengxin Liu <mengxin@alauda.io>
Date: Sun, 30 Mar 2025 06:15:20 +0000
Subject: [PATCH] increase election timer

---
 ovsdb/raft.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/ovsdb/raft.c b/ovsdb/raft.c
index ac3d37ac409..99fad50fe82 100644
--- a/ovsdb/raft.c
+++ b/ovsdb/raft.c
@@ -204,7 +204,7 @@ struct raft {
      * most recently added to 'log'. */
     struct hmap servers;
 
-#define ELECTION_BASE_MSEC 1000
+#define ELECTION_BASE_MSEC 5000
 #define ELECTION_RANGE_MSEC 1000
 #define ELECTION_MIN_MSEC 100
 #define ELECTION_MAX_MSEC 600000
