diff --git a/controller/ovn-controller.c b/controller/ovn-controller.c
index 542619aee..3c5aa7773 100644
--- a/controller/ovn-controller.c
+++ b/controller/ovn-controller.c
@@ -25,0 +26 @@
+#include <sys/utsname.h>
@@ -107,0 +109,2 @@ VLOG_DEFINE_THIS_MODULE(main);
+bool lflow_kube_ovn_skip_ct = true;
+
@@ -7381,0 +7385,11 @@ main(int argc, char *argv[])
+    struct utsname info;
+    if (!uname(&info)) {
+        int kernel, major, minor, patch;
+        if (strstr(info.release, "el8") &&
+            sscanf(info.release, "%d.%d.%d-%d", &kernel, &major,
+                   &minor, &patch) == 4 && kernel == 4 && major == 18 &&
+            minor == 0 && patch == 372) {
+            lflow_kube_ovn_skip_ct = false;
+        }
+    }
+
diff --git a/northd/lflow-mgr.c b/northd/lflow-mgr.c
index 7e7d0b1bd..f368e7e3e 100644
--- a/northd/lflow-mgr.c
+++ b/northd/lflow-mgr.c
@@ -43 +43,2 @@ static void ovn_lflow_init(struct ovn_lflow *,
-                           const char *flow_desc, struct uuid sbuuid);
+                           const char *flow_desc, struct uuid sbuuid,
+                           const char *kube_ovn_hint);
@@ -62 +63 @@ static struct ovn_lflow *do_ovn_lflow_add(
-    bool acl_ct_translation);
+    bool acl_ct_translation, const char *kube_ovn_hint);
@@ -189,0 +191 @@ struct ovn_lflow {
+    char *kube_ovn_hint;
@@ -744 +746,2 @@ lflow_table_add_lflow__(struct lflow_table *lflow_table,
-                        struct lflow_ref *lflow_ref)
+                        struct lflow_ref *lflow_ref,
+                        const char *kube_ovn_hint)
@@ -766 +769 @@ lflow_table_add_lflow__(struct lflow_table *lflow_table,
-                         acl_ct_translation);
+                         acl_ct_translation, kube_ovn_hint);
@@ -827 +830 @@ lflow_table_add_lflow(struct lflow_table_add_args *args)
-                            args->lflow_ref);
+                            args->lflow_ref, args->kube_ovn_hint);
@@ -945 +948,2 @@ ovn_lflow_init(struct ovn_lflow *lflow,
-               const char *where, const char *flow_desc, struct uuid sbuuid)
+               const char *where, const char *flow_desc, struct uuid sbuuid,
+               const char *kube_ovn_hint)
@@ -961,0 +966 @@ ovn_lflow_init(struct ovn_lflow *lflow,
+    lflow->kube_ovn_hint = nullable_xstrdup(kube_ovn_hint);
@@ -1040,0 +1046 @@ ovn_lflow_destroy(struct lflow_table *lflow_table, struct ovn_lflow *lflow)
+    free(lflow->kube_ovn_hint);
@@ -1056 +1062 @@ do_ovn_lflow_add(struct lflow_table *lflow_table, size_t dp_bitmap_len,
-                 bool acl_ct_translation)
+                 bool acl_ct_translation, const char *kube_ovn_hint)
@@ -1087 +1093 @@ do_ovn_lflow_add(struct lflow_table *lflow_table, size_t dp_bitmap_len,
-                   flow_desc, sbuuid);
+                   flow_desc, sbuuid, kube_ovn_hint);
@@ -1176,0 +1183,3 @@ sync_lflow_to_sb(struct ovn_lflow *lflow,
+        if (lflow->kube_ovn_hint) {
+            smap_add(&ids, "kube-ovn-hint", lflow->kube_ovn_hint);
+        }
@@ -1190,0 +1200,2 @@ sync_lflow_to_sb(struct ovn_lflow *lflow,
+            const char *kube_ovn_hint = smap_get_def(
+                &sbflow->external_ids, "kube-ovn-hint", "");
@@ -1220,0 +1232,5 @@ sync_lflow_to_sb(struct ovn_lflow *lflow,
+            if (lflow->kube_ovn_hint &&
+                strcmp(kube_ovn_hint, lflow->kube_ovn_hint)) {
+                sbrec_logical_flow_update_external_ids_setkey(
+                    sbflow, "kube-ovn-hint", lflow->kube_ovn_hint);
+            }
diff --git a/northd/lflow-mgr.h b/northd/lflow-mgr.h
index 84d0b3e67..f3ada189a 100644
--- a/northd/lflow-mgr.h
+++ b/northd/lflow-mgr.h
@@ -91,0 +92 @@ struct lflow_table_add_args {
+    const char *kube_ovn_hint;
@@ -110,0 +112 @@ void lflow_table_add_lflow(struct lflow_table_add_args *args);
+#define WITH_KUBE_OVN_HINT(HINT) .kube_ovn_hint = HINT
@@ -132,0 +135,25 @@ void lflow_table_add_lflow(struct lflow_table_add_args *args);
+#define ovn_lflow_add_with_hint(LFLOW_TABLE, OD, STAGE, PRIORITY, MATCH, \
+                                ACTIONS, STAGE_HINT, LFLOW_REF) \
+    ovn_lflow_add(LFLOW_TABLE, OD, STAGE, PRIORITY, MATCH, ACTIONS, \
+                  LFLOW_REF, WITH_HINT(STAGE_HINT))
+
+#define ovn_lflow_add_with_lport_and_hint(LFLOW_TABLE, OD, STAGE, PRIORITY, \
+                                          MATCH, ACTIONS, IN_OUT_PORT, \
+                                          STAGE_HINT, LFLOW_REF) \
+    ovn_lflow_add(LFLOW_TABLE, OD, STAGE, PRIORITY, MATCH, ACTIONS, \
+                  LFLOW_REF, WITH_IO_PORT(IN_OUT_PORT), WITH_HINT(STAGE_HINT))
+
+#define ovn_lflow_add_with_hint__(LFLOW_TABLE, OD, STAGE, PRIORITY, MATCH, \
+                                  ACTIONS, IN_OUT_PORT, CTRL_METER, \
+                                  STAGE_HINT, LFLOW_REF) \
+    ovn_lflow_add(LFLOW_TABLE, OD, STAGE, PRIORITY, MATCH, ACTIONS, \
+                  LFLOW_REF, WITH_IO_PORT(IN_OUT_PORT), \
+                  WITH_CTRL_METER(CTRL_METER), WITH_HINT(STAGE_HINT))
+
+#define ovn_lflow_add_with_kube_ovn_hint(LFLOW_TABLE, OD, STAGE, PRIORITY, \
+                                         MATCH, ACTIONS, STAGE_HINT, \
+                                         LFLOW_REF) \
+    ovn_lflow_add(LFLOW_TABLE, OD, STAGE, PRIORITY, MATCH, ACTIONS, \
+                  LFLOW_REF, WITH_HINT(STAGE_HINT), \
+                  WITH_KUBE_OVN_HINT(OVN_LFLOW_HINT_KUBE_OVN_SKIP_CT))
+
diff --git a/northd/northd.c b/northd/northd.c
index 30fe27787..51d32968e 100644
--- a/northd/northd.c
+++ b/northd/northd.c
@@ -9221 +9222 @@ build_lswitch_mod_dl_src_rules(struct ovn_port *op,
-        !op->od->n_router_ports || !op->od->n_localnet_ports) {
+        vector_is_empty(&op->od->router_ports) || vector_is_empty(&op->od->localnet_ports)) {
@@ -9246,2 +9247,3 @@ build_lswitch_mod_dl_src_rules(struct ovn_port *op,
-    for (size_t i = 0; i < op->od->n_router_ports; i++) {
-        struct ovn_port *rp = op->od->router_ports[i];
+    for (size_t i = 0; i < vector_len(&op->od->router_ports); i++) {
+        struct ovn_port *rp = vector_get(&op->od->router_ports, i,
+                                         struct ovn_port *);
@@ -10758,2 +10760,3 @@ build_lswitch_arp_nd_responder_known_ips(struct ovn_port *op,
-            for (size_t i = 0; i < op->od->n_localnet_ports; i++) {
-                struct ovn_port *localnet_port = op->od->localnet_ports[i];
+            for (size_t i = 0; i < vector_len(&op->od->localnet_ports); i++) {
+                struct ovn_port *localnet_port = vector_get(
+                    &op->od->localnet_ports, i, struct ovn_port *);
