diff --git a/ovsdb/ovsdb-tool.c b/ovsdb/ovsdb-tool.c
index ea73150c6e..4c8bfcb74e 100644
--- a/ovsdb/ovsdb-tool.c
+++ b/ovsdb/ovsdb-tool.c
@@ -2111 +2111 @@ do_fix_cluster(struct ovs_cmdl_context *ctx)
-        check_ovsdb_error(ovsdb_log_replace(s->log, c.entries[i]->array.elems, c.n_entries[i]));
+        check_ovsdb_error(ovsdb_log_replace(s->log, c.entries[i]->array.elements, c.n_entries[i]));
